enhancements/cloud-integration/openstack/cloud-provider-config-upgrade.md

```diff
@@ -0,0 +1,420 @@
+---
+title: OpenStack Cloud Provider Config Upgrade
+authors:
+  - "@stephenfin"
+  - "@mdbooth"
+reviewers:
+  - TBD
+approvers:
+  - "@jspeed"
+  - TBD
+api-approvers: # in case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers)
+creation-date: 2022-01-18
+last-updated: 2022-01-25
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - https://issues.redhat.com/browse/OSASINFRA-2758
+see-also:
+replaces:
+superseded-by:
+---
+
+# OpenStack Cloud Provider Config Upgrade
+
+## Summary
+
+Cloud configuration is stored in the user-managed configmap
+`cloud-provider-config` in `openshift-config`.  In 4.11 we want to switch
+OpenStack clusters from the legacy cloud provider to the external Cloud
+Controller Manager (CCM). The configuration for the legacy and external cloud
+providers are similar, but not identical.
+
+This enhancement will describe how we can ensure correct configuration is passed
+to the cloud provider in use during and after an upgrade, and how we can prevent
+an upgrade to 4.11 if the configuration cannot be upgraded.
+
+## Motivation
+
+The legacy cloud provider is dead. Long live the external cloud provider.
+
+### Goals
+
+- No user action is required in the common case
+- Upgrade legacy cloud provider to external cloud controller manager where possible
+- Mark cluster as non-upgradable where upgrade to external CCM is not possible
+
+### Non-Goals
+
+- Modify user-managed configuration
+- Manage cloud providers for other, non-OpenStack platforms
+
+## Proposal
+
+- The change will be implemented in the Cluster Config Operator (CCO)
+- CCO will create a separate configmap for the external cloud provider
+- We will prevent upgrade if it is not possible to create the external cloud provider configmap
+
+### Implemented in the Cluster Config Operator (CCO)
+
+The new functionality will be implemented as a change to CCO. CCO already
+applies a transformation function to the user-managed config and writes the
+result to `openshift-config-managed/kube-cloud-config`. The transformation
+function is selected according to PlatformType in the cluster Infrastructure
+object. Components requiring legacy cloud provider configuration, including
+kubelet and KCM, read the transformed configmap.
+
+### Separate ConfigMaps for legacy and external cloud providers
+
+The cloud configuration is read by multiple components which will not all be
+upgraded simultaneously. If we simply updated the `kube-cloud-config` to be
+compatible with external CCM we would risk breaking KCM and kubelets which
+have not yet been upgraded. Therefore we are constrained to write two versions
+of the managed config map.
+
+CCO will be modified to create both `kube-cloud-config` and
+`cloud-controller-manager-config`. `kube-cloud-config` will remain compatible
```

Comment:
this is  odd. Why isn't this done by the cloud-controller-magner-operator?

Reply:
For the record, we discussed this offline. The new plan is to:
* Leave CCO alone. It will continue to manage the legacy config because it's required by kubelet and KCM.
* Move most of the behaviour described in this enhancement to CCCMO.
* Remove the legacy behaviour from CCO in 4.12

We'll update the enhancement to reflect the new plan.

Reply:
aaand updated